File: unicalib_C_plus_plus/src/click_calib_wrapper.cpp
/**
 * UniCalib C++ — click_calib BA：自动关键点、写 YAML、调 optimize.py、解析结果
 */
#include "unicalib/click_calib_wrapper.hpp"
#include "unicalib/process_runner.hpp"
#include "unicalib/logger.hpp"
#include <filesystem>
#include <fstream>
#include <algorithm>
#include <cmath>

#if defined(UNICALIB_USE_YAML) && UNICALIB_USE_YAML
#include <yaml-cpp/yaml.h>
#endif

#if defined(UNICALIB_USE_OPENCV) && UNICALIB_USE_OPENCV
#include <opencv2/features2d.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#endif

namespace fs = std::filesystem;
namespace unicalib {

namespace {

constexpr double kSyncThresholdSec = 0.1;
constexpr int kMinKeypoints = 10;
constexpr int kMaxFramesForKeypoints = 5;

}  // namespace

std::optional<CalibResult> run_click_calib_ba(
  DataManager& data_mgr,
  const SensorConfig& sensor_a,
  const SensorConfig& sensor_b,
  const std::unordered_map<std::string, IntrinsicResultHolder>& intrinsics,
  const std::optional<CalibResult>& initial,
  const ExternalToolsConfig& tools_config,
  int timeout_seconds) {

  if (tools_config.click_calib.empty()) {
    LOG_WARNING("click_calib not configured for Camera-Camera bundle adjustment.");
    LOG_WARNING("Set 'third_party.click_calib' in config.yaml or set UNICALIB_CLICK_CALIB environment variable.");
    LOG_WARNING("Example config:");
    LOG_WARNING("  third_party:");
    LOG_WARNING("    click_calib: \"/path/to/click_calib\"");
    LOG_WARNING("Example environment variable:");
    LOG_WARNING("  export UNICALIB_CLICK_CALIB=/path/to/click_calib");
    return std::nullopt;
  }

  auto it_a = intrinsics.find(sensor_a.sensor_id);
  auto it_b = intrinsics.find(sensor_b.sensor_id);
  if (it_a == intrinsics.end() || it_b == intrinsics.end()) {
    LOG_WARNING("click_calib: missing intrinsics for " + sensor_a.sensor_id + " or " + sensor_b.sensor_id);
    return std::nullopt;
  }

#if !(defined(UNICALIB_USE_OPENCV) && UNICALIB_USE_OPENCV) || !(defined(UNICALIB_USE_YAML) && UNICALIB_USE_YAML)
  LOG_WARNING("click_calib BA requires OpenCV and yaml-cpp");
  return std::nullopt;
#else

  // --- 收集同步图像路径 ---
  std::vector<std::pair<double, std::string>> paths_a, paths_b;
  data_mgr.iter_image_paths(sensor_a.topic, [&](double ts, const std::string& path) {
    paths_a.emplace_back(ts, path);
  }, (kMaxFramesForKeypoints + 2) * 2, 1);
  data_mgr.iter_image_paths(sensor_b.topic, [&](double ts, const std::string& path) {
    paths_b.emplace_back(ts, path);
  }, (kMaxFramesForKeypoints + 2) * 2, 1);
  if (paths_a.empty() || paths_b.empty()) {
    LOG_WARNING("click_calib: no image paths for " + sensor_a.topic + " or " + sensor_b.topic);
    return std::nullopt;
  }
  std::sort(paths_a.begin(), paths_a.end());
  std::sort(paths_b.begin(), paths_b.end());

  // --- SIFT 提取匹配点对（多帧聚合）---
  std::vector<std::pair<double, double>> keypoints_a, keypoints_b;  // (u,v) per point

  cv::Ptr<cv::SIFT> sift = cv::SIFT::create(200);
  int frames_used = 0;
  for (const auto& [ts_a, path_a] : paths_a) {
    if (frames_used >= kMaxFramesForKeypoints) break;
    auto it = std::lower_bound(paths_b.begin(), paths_b.end(), ts_a - kSyncThresholdSec,
        [](const std::pair<double, std::string>& x, double t) { return x.first < t; });
    if (it == paths_b.end() || std::abs(it->first - ts_a) > kSyncThresholdSec) continue;

    cv::Mat img_a = cv::imread(path_a);
    cv::Mat img_b = cv::imread(it->second);
    if (img_a.empty() || img_b.empty()) continue;

    cv::Mat gray_a, gray_b;
    cv::cvtColor(img_a, gray_a, cv::COLOR_BGR2GRAY);
    cv::cvtColor(img_b, gray_b, cv::COLOR_BGR2GRAY);

    std::vector<cv::KeyPoint> kp_a, kp_b;
    cv::Mat desc_a, desc_b;
    sift->detectAndCompute(gray_a, cv::noArray(), kp_a, desc_a);
    sift->detectAndCompute(gray_b, cv::noArray(), kp_b, desc_b);
    if (desc_a.empty() || desc_b.empty()) continue;

    cv::BFMatcher matcher(cv::NORM_L2);
    std::vector<cv::DMatch> matches;
    matcher.match(desc_a, desc_b, matches);
    for (const auto& m : matches) {
      const auto& pa = kp_a[m.queryIdx].pt;
      const auto& pb = kp_b[m.trainIdx].pt;
      keypoints_a.emplace_back(static_cast<double>(pa.x), static_cast<double>(pa.y));
      keypoints_b.emplace_back(static_cast<double>(pb.x), static_cast<double>(pb.y));
    }
    ++frames_used;
  }

  if (keypoints_a.size() < static_cast<size_t>(kMinKeypoints)) {
    LOG_WARNING("click_calib: insufficient keypoints (" + std::to_string(keypoints_a.size()) + " < " + std::to_string(kMinKeypoints) + ")");
    return std::nullopt;
  }

  std::string tmpdir = (fs::temp_directory_path() / "unicalib_click_calib").string();
  try { fs::create_directories(tmpdir); } catch (...) {
    LOG_ERROR("click_calib: failed to create " + tmpdir);
    return std::nullopt;
  }

  std::string calib_path = tmpdir + "/calib_init.yaml";
  std::string kp_path = tmpdir + "/keypoints.yaml";
  std::string output_path = tmpdir + "/result.yaml";

  // --- 写 calib_init.yaml ---
  YAML::Emitter calib_out;
  calib_out << YAML::BeginMap;
  calib_out << YAML::Key << "cameras" << YAML::Value << YAML::BeginSeq;

  for (const auto* it : {&it_a->second, &it_b->second}) {
    const std::string& sid = (it == &it_a->second) ? sensor_a.sensor_id : sensor_b.sensor_id;
    calib_out << YAML::BeginMap;
    calib_out << YAML::Key << "camera_id" << YAML::Value << sid;
    if (it->camera_pinhole) {
      const auto& K = it->camera_pinhole->K;
      const auto& d = it->camera_pinhole->dist_coeffs;
      calib_out << YAML::Key << "intrinsic_type" << YAML::Value << "pinhole";
      calib_out << YAML::Key << "fx" << YAML::Value << K(0, 0);
      calib_out << YAML::Key << "fy" << YAML::Value << K(1, 1);
      calib_out << YAML::Key << "cx" << YAML::Value << K(0, 2);
      calib_out << YAML::Key << "cy" << YAML::Value << K(1, 2);
      calib_out << YAML::Key << "k1" << YAML::Value << (d.size() > 0 ? d[0] : 0.0);
      calib_out << YAML::Key << "k2" << YAML::Value << (d.size() > 1 ? d[1] : 0.0);
      calib_out << YAML::Key << "p1" << YAML::Value << (d.size() > 2 ? d[2] : 0.0);
      calib_out << YAML::Key << "p2" << YAML::Value << (d.size() > 3 ? d[3] : 0.0);
    } else if (it->camera_fisheye) {
      calib_out << YAML::Key << "intrinsic_type" << YAML::Value << it->camera_fisheye->model_type;
      for (const auto& [k, v] : it->camera_fisheye->params)
        calib_out << YAML::Key << k << YAML::Value << v;
    }
    calib_out << YAML::EndMap;
  }
  calib_out << YAML::EndSeq;

  if (initial) {
    calib_out << YAML::Key << "initial_extrinsic" << YAML::Value << YAML::BeginMap;
    calib_out << YAML::Key << "from" << YAML::Value << initial->pair.first;
    calib_out << YAML::Key << "to" << YAML::Value << initial->pair.second;
    calib_out << YAML::Key << "rotation" << YAML::Value << YAML::BeginSeq;
    for (int i = 0; i < 3; ++i)
      calib_out << YAML::Flow << YAML::BeginSeq << initial->rotation(i,0) << initial->rotation(i,1) << initial->rotation(i,2) << YAML::EndSeq;
    calib_out << YAML::EndSeq;
    calib_out << YAML::Key << "translation" << YAML::Value << YAML::Flow << YAML::BeginSeq
              << initial->translation(0) << initial->translation(1) << initial->translation(2) << YAML::EndSeq;
    calib_out << YAML::EndMap;
  }
  calib_out << YAML::EndMap;

  std::ofstream f_calib(calib_path);
  if (!f_calib) { LOG_ERROR("click_calib: cannot write " + calib_path); return std::nullopt; }
  f_calib << calib_out.c_str();

  // --- 写 keypoints.yaml: camera_a / camera_b 为 [[u,v],...] ---
  YAML::Emitter kp_out;
  kp_out << YAML::BeginMap;
  kp_out << YAML::Key << "camera_a" << YAML::Value << YAML::BeginSeq;
  for (const auto& p : keypoints_a)
    kp_out << YAML::Flow << YAML::BeginSeq << p.first << p.second << YAML::EndSeq;
  kp_out << YAML::EndSeq;
  kp_out << YAML::Key << "camera_b" << YAML::Value << YAML::BeginSeq;
  for (const auto& p : keypoints_b)
    kp_out << YAML::Flow << YAML::BeginSeq << p.first << p.second << YAML::EndSeq;
  kp_out << YAML::EndSeq << YAML::EndMap;

  std::ofstream f_kp(kp_path);
  if (!f_kp) { LOG_ERROR("click_calib: cannot write " + kp_path); return std::nullopt; }
  f_kp << kp_out.c_str();

  // --- 调用 python source/optimize.py ---
  std::string script = tools_config.click_calib + "/source/optimize.py";
  if (!fs::exists(script)) {
    LOG_WARNING("click_calib: optimize.py not found at " + script);
    return std::nullopt;
  }
  std::string python = find_python3();
  if (python.empty()) { LOG_WARNING("click_calib: no python3"); return std::nullopt; }

  std::unordered_map<std::string, std::string> env;
  env["PYTHONPATH"] = tools_config.click_calib + "/source";

  ProcessResult res = run_process(
    { python, script, "--calib", calib_path, "--keypoints", kp_path, "--output", output_path },
    env, timeout_seconds, "");

  if (res.exit_code != 0 || res.timed_out) {
    LOG_WARNING("click_calib optimize exit_code=" + std::to_string(res.exit_code));
    return std::nullopt;
  }

  if (!fs::exists(output_path)) {
    LOG_WARNING("click_calib: result not found " + output_path);
    return std::nullopt;
  }

  YAML::Node data;
  try {
    data = YAML::LoadFile(output_path);
  } catch (const std::exception& e) {
    LOG_ERROR(std::string("click_calib result parse: ") + e.what());
    return std::nullopt;
  }

  CalibResult cr;
  cr.pair = {sensor_a.sensor_id, sensor_b.sensor_id};
  cr.rotation = Eigen::Matrix3d::Identity();
  cr.translation = Eigen::Vector3d::Zero();
  if (data["rotation"]) {
    const auto& Rnode = data["rotation"];
    for (int i = 0; i < 3 && i < static_cast<int>(Rnode.size()); ++i)
      for (int j = 0; j < 3 && j < static_cast<int>(Rnode[i].size()); ++j)
        cr.rotation(i, j) = Rnode[i][j].as<double>();
  }
  if (data["translation"]) {
    const auto& tnode = data["translation"];
    for (size_t i = 0; i < 3 && i < tnode.size(); ++i)
      cr.translation(i) = tnode[i].as<double>();
  }
  if (data["mde"]) cr.reprojection_error = data["mde"].as<double>();
  cr.confidence = 0.8;
  cr.method_used = "click_calib_BA";
  LOG_INFO("click_calib BA result " + cr.pair.first + "->" + cr.pair.second + " mde=" + std::to_string(cr.reprojection_error));
  return cr;
#endif
}

}  // namespace unicalib

File: UniCalib/unicalib/validation/click_calib_wrapper.py
"""
click_calib 封装器 —— 环视相机 Camera-Camera 全局 BA 精标定
项目位置: /root/calib_ws/src/click_calib/
论文: Click-Calib: A Robust Extrinsic Calibration Method for Surround-View Systems
"""
from __future__ import annotations
import logging
import os
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Dict, Optional

import cv2
import numpy as np
import yaml
from scipy.spatial.transform import Rotation

from ..core.calib_result import CalibResult, CameraIntrinsic, FisheyeIntrinsic
from ..core.sensor_config import SensorConfig
from ..core.data_manager import DataManager

logger = logging.getLogger(__name__)

WORKSPACE = os.environ.get("WORKSPACE", "/root/calib_ws")
CLICK_CALIB_PATH = os.path.join(WORKSPACE, "src", "click_calib")


class ClickCalibWrapper:
    """
    click_calib 封装器，用于:
      1. Camera-Camera 外参全局 BA 精标定
      2. 人工验证界面启动
      3. BEV 图像生成（用于可视化验证）

    click_calib 原本需要人工选点，这里通过自动特征匹配代替人工操作。
    """

    def __init__(self, system_config: dict):
        self._available = Path(CLICK_CALIB_PATH).exists()
        if not self._available:
            logger.warning(f"click_calib not found at {CLICK_CALIB_PATH}")

    def global_ba(
        self,
        data_mgr: DataManager,
        sensor_a: SensorConfig,
        sensor_b: SensorConfig,
        intrinsics: Dict,
        initial: Optional[CalibResult],
    ) -> CalibResult:
        """
        执行 Camera-Camera 全局 BA 精标定。
        自动提取特征点替代 click_calib 的人工选点步骤。
        """
        if self._available:
            result = self._run_click_calib_optimize(
                data_mgr, sensor_a, sensor_b, intrinsics, initial)
            if result is not None:
                return result

        # Fallback: 直接用特征匹配精化
        return self._feature_ba_fallback(
            data_mgr, sensor_a, sensor_b, intrinsics, initial)

    def launch_verify_gui(
        self, data_path: str,
        sensors: Dict[str, SensorConfig],
        intrinsics: Dict,
        extrinsics: Dict,
    ):
        """启动 click_calib 人工验证 GUI（非阻塞）。"""
        if not self._available:
            logger.warning("click_calib not available for GUI verification.")
            return

        # 准备验证数据文件
        verify_data = {
            "intrinsics": {
                k: (v.to_dict() if isinstance(v, CameraIntrinsic)
                    else v.to_dict() if isinstance(v, FisheyeIntrinsic)
                    else {})
                for k, v in intrinsics.items()
            },
            "extrinsics": {
                f"{k[0]}_to_{k[1]}": {
                    "rotation": v.rotation.tolist(),
                    "translation": v.translation.tolist(),
                }
                for k, v in extrinsics.items()
            },
        }
        verify_yaml = "/tmp/unicalib_verify.yaml"
        with open(verify_yaml, "w") as f:
            yaml.dump(verify_data, f)

        try:
            subprocess.Popen([
                sys.executable,
                os.path.join(CLICK_CALIB_PATH, "source", "projection.py"),
                "--calib", verify_yaml,
                "--data", data_path,
                "--mode", "verify",
            ])
            logger.info("click_calib verification GUI launched.")
        except Exception as e:
            logger.warning(f"Failed to launch click_calib GUI: {e}")

    # ------------------------------------------------------------------
    # click_calib optimize.py 调用
    # ------------------------------------------------------------------

    def _run_click_calib_optimize(
        self,
        data_mgr: DataManager,
        sensor_a: SensorConfig,
        sensor_b: SensorConfig,
        intrinsics: Dict,
        initial: Optional[CalibResult],
    ) -> Optional[CalibResult]:
        """
        调用 click_calib 的 optimize.py 进行 BA 优化。
        自动生成对应关键点（替代手动点击）。
        """
        try:
            intr_a = intrinsics.get(sensor_a.sensor_id)
            intr_b = intrinsics.get(sensor_b.sensor_id)
            if intr_a is None or intr_b is None:
                return None

            with tempfile.TemporaryDirectory() as tmpdir:
                # 提取自动特征点
                keypoints_a, keypoints_b = self._auto_keypoints(
                    data_mgr, sensor_a, sensor_b, intr_a, intr_b)

                if len(keypoints_a) < 10:
                    logger.warning("Insufficient keypoints for click_calib BA.")
                    return None

                # 构建 click_calib 格式的校准文件
                calib_data = self._build_click_calib_format(
                    sensor_a, sensor_b, intr_a, intr_b, initial)
                calib_path = os.path.join(tmpdir, "calib_init.yaml")
                with open(calib_path, "w") as f:
                    yaml.dump(calib_data, f)

                # 构建关键点文件
                kp_path = os.path.join(tmpdir, "keypoints.yaml")
                with open(kp_path, "w") as f:
                    yaml.dump({
                        "camera_a": keypoints_a.tolist(),
                        "camera_b": keypoints_b.tolist(),
                    }, f)

                output_path = os.path.join(tmpdir, "result.yaml")

                env = os.environ.copy()
                env["PYTHONPATH"] = (
                    f"{CLICK_CALIB_PATH}/source:"
                    f"{env.get('PYTHONPATH', '')}"
                )
                proc = subprocess.run(
                    [
                        sys.executable,
                        os.path.join(CLICK_CALIB_PATH, "source", "optimize.py"),
                        "--calib", calib_path,
                        "--keypoints", kp_path,
                        "--output", output_path,
                    ],
                    env=env, capture_output=True, text=True, timeout=120
                )

                if proc.returncode == 0 and Path(output_path).exists():
                    return self._parse_click_result(
                        output_path, sensor_a.sensor_id, sensor_b.sensor_id)

        except Exception as e:
            logger.warning(f"click_calib optimize failed: {e}")

        return None

    def _auto_keypoints(
        self, data_mgr, sensor_a, sensor_b, intr_a, intr_b
    ):
        """从同步图像对中自动提取对应关键点。"""
        pairs = list(data_mgr.iter_synced_frames(
            sensor_a.topic, sensor_b.topic, max_pairs=5))

        all_a, all_b = [], []
        sift = cv2.SIFT_create(nfeatures=200) if hasattr(cv2, 'SIFT_create') else None
        if sift is None:
            return np.zeros((0, 2)), np.zeros((0, 2))

        for _, img_a, img_b in pairs:
            gray_a = cv2.cvtColor(img_a, cv2.COLOR_BGR2GRAY)
            gray_b = cv2.cvtColor(img_b, cv2.COLOR_BGR2GRAY)
            kp_a, desc_a = sift.detectAndCompute(gray_a, None)
            kp_b, desc_b = sift.detectAndCompute(gray_b, None)

            if desc_a is None or desc_b is None:
                continue

            matcher = cv2.BFMatcher(cv2.NORM_L2)
            raw = matcher.knnMatch(desc_a, desc_b, k=2)
            good = [m for m, n in raw if m.distance < 0.7 * n.distance]

            for m in good[:20]:
                all_a.append(kp_a[m.queryIdx].pt)
                all_b.append(kp_b[m.trainIdx].pt)

        return np.array(all_a), np.array(all_b)

    def _build_click_calib_format(self, sa, sb, intr_a, intr_b, initial):
        """构建 click_calib 格式的标定文件。"""
        def _intr_dict(intr, sid):
            if isinstance(intr, CameraIntrinsic):
                return {
                    "camera_id": sid,
                    "intrinsic_type": "pinhole",
                    "fx": float(intr.K[0, 0]),
                    "fy": float(intr.K[1, 1]),
                    "cx": float(intr.K[0, 2]),
                    "cy": float(intr.K[1, 2]),
                    "k1": float(intr.dist_coeffs[0]),
                    "k2": float(intr.dist_coeffs[1]) if len(intr.dist_coeffs) > 1 else 0.0,
                    "p1": float(intr.dist_coeffs[2]) if len(intr.dist_coeffs) > 2 else 0.0,
                    "p2": float(intr.dist_coeffs[3]) if len(intr.dist_coeffs) > 3 else 0.0,
                }
            elif isinstance(intr, FisheyeIntrinsic):
                return {"camera_id": sid, "intrinsic_type": intr.model_type, **intr.params}
            return {"camera_id": sid}

        data = {
            "cameras": [
                _intr_dict(intr_a, sa.sensor_id),
                _intr_dict(intr_b, sb.sensor_id),
            ],
        }
        if initial:
            data["initial_extrinsic"] = {
                "from": sa.sensor_id, "to": sb.sensor_id,
                "rotation": initial.rotation.tolist(),
                "translation": initial.translation.tolist(),
            }
        return data

    def _parse_click_result(self, result_yaml: str, sa_id: str, sb_id: str) -> CalibResult:
        with open(result_yaml) as f:
            data = yaml.safe_load(f)
        R = np.array(data.get("rotation", np.eye(3).tolist()))
        t = np.array(data.get("translation", [0, 0, 0]))
        return CalibResult(
            pair=(sa_id, sb_id),
            rotation=R, translation=t.flatten(),
            reprojection_error=float(data.get("mde", float('inf'))),
            confidence=0.8,
            method_used="click_calib_BA",
        )

    # ------------------------------------------------------------------
    # Fallback: 特征点 BA
    # ------------------------------------------------------------------

    def _feature_ba_fallback(
        self, data_mgr, sensor_a, sensor_b, intrinsics, initial
    ) -> CalibResult:
        """基于特征点的简化 BA 优化。"""
        if initial is None:
            return CalibResult(
                pair=(sensor_a.sensor_id, sensor_b.sensor_id),
                rotation=np.eye(3), translation=np.zeros(3),
                confidence=0.0, method_used="identity_fallback")

        # 简单返回初始值 (精化需要第三方工具)
        return CalibResult(
            pair=(sensor_a.sensor_id, sensor_b.sensor_id),
            rotation=initial.rotation.copy(),
            translation=initial.translation.copy(),
            confidence=initial.confidence,
            method_used=f"{initial.method_used}+feature_ba_fallback",
        )
